File: frontend/src/lib/api/analysis.ts
import { api } from './axios-instance'

// === Escalation Types ===

export interface EscalationTrendPoint {
  date: string
  high_count: number
  medium_count: number
  low_count: number
  avg_score: number
}

export interface EscalationTrendResponse {
  period: string
  trend: EscalationTrendPoint[]
  total_high: number
  total_medium: number
  total_low: number
}

// === Correlation Types ===

export interface CorrelationResponse {
  id: string
  duplicate_group_id: string
  analysis_text: string | null
  consistent_facts: string[]
  unique_details: string[]
  contradictions: string[]
  source_count: number
  created_at: string
}

export interface CorrelationListResponse {
  correlations: CorrelationResponse[]
  total: number
}

// === Pattern Types ===

export interface DetectedPatternResponse {
  id: string
  pattern_type: string
  title: string
  description: string | null
  evidence_message_ids: string[]
  confidence: number
  detected_at: string
  expires_at: string | null
}

export interface PatternListResponse {
  patterns: DetectedPatternResponse[]
  total: number
}

// === Timeline Types ===

export interface TimelineEvent {
  date: string
  description: string
  sources: string[]
  significance: number
}

export interface TimelineResponse {
  id: string
  user_id: string
  title: string
  topic: string | null
  collection_id: string | null
  events: TimelineEvent[]
  date_range_start: string | null
  date_range_end: string | null
  message_count: number
  created_at: string
}

export interface TimelineListResponse {
  timelines: TimelineResponse[]
  total: number
}

export interface TimelineGenerateRequest {
  topic?: string
  collection_id?: string
  start_date?: string
  end_date?: string
}

// === API Client ===

export const analysisApi = {
  // Escalation
  getEscalationTrend: (params?: { collection_id?: string; period?: string }) =>
    api.get<EscalationTrendResponse>('/api/analysis/escalation', { params }),

  // Correlations
  listCorrelations: (limit = 10) =>
    api.get<CorrelationListResponse>('/api/analysis/correlations', { params: { limit } }),

  getCorrelation: (duplicateGroupId: string) =>
    api.get<CorrelationResponse>(`/api/analysis/correlations/${duplicateGroupId}`),

  // Patterns
  listPatterns: (params?: { period?: string; limit?: number }) =>
    api.get<PatternListResponse>('/api/analysis/patterns', { params }),

  getPattern: (patternId: string) =>
    api.get<DetectedPatternResponse>(`/api/analysis/patterns/${patternId}`),

  // Timelines
  generateTimeline: (request: TimelineGenerateRequest) =>
    api.post<TimelineResponse>('/api/analysis/timeline/generate', request),

  listTimelines: (limit = 20) =>
    api.get<TimelineListResponse>('/api/analysis/timelines', { params: { limit } }),

  getTimeline: (timelineId: string) =>
    api.get<TimelineResponse>(`/api/analysis/timelines/${timelineId}`),

  deleteTimeline: (timelineId: string) =>
    api.delete(`/api/analysis/timelines/${timelineId}`),
}

File: backend/app/schemas/analysis.py
"""Pydantic schemas for intelligence analysis features."""

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# === Escalation Scoring ===

class EscalationScoreResponse(BaseModel):
    id: UUID
    message_id: UUID
    score: float
    level: str
    factors: List[str] = []
    computed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class EscalationTrendPoint(BaseModel):
    date: str
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    avg_score: float = 0.0


class EscalationTrendResponse(BaseModel):
    period: str
    trend: List[EscalationTrendPoint] = []
    total_high: int = 0
    total_medium: int = 0
    total_low: int = 0


# === Cross-Source Correlation ===

class CorrelationResponse(BaseModel):
    id: UUID
    duplicate_group_id: UUID
    analysis_text: Optional[str] = None
    consistent_facts: List[str] = []
    unique_details: List[str] = []
    contradictions: List[str] = []
    source_count: int = 0
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CorrelationListResponse(BaseModel):
    correlations: List[CorrelationResponse] = []
    total: int = 0


# === Pattern Detection ===

class DetectedPatternResponse(BaseModel):
    id: UUID
    pattern_type: str
    title: str
    description: Optional[str] = None
    evidence_message_ids: List[str] = []
    confidence: float = 0.5
    detected_at: datetime
    expires_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class PatternListResponse(BaseModel):
    patterns: List[DetectedPatternResponse] = []
    total: int = 0


# === Timeline Reconstruction ===

class TimelineEvent(BaseModel):
    date: str
    description: str
    sources: List[str] = []
    significance: int = 1  # 1-5


class TimelineResponse(BaseModel):
    id: UUID
    user_id: UUID
    title: str
    topic: Optional[str] = None
    collection_id: Optional[UUID] = None
    events: List[TimelineEvent] = []
    date_range_start: Optional[datetime] = None
    date_range_end: Optional[datetime] = None
    message_count: int = 0
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TimelineListResponse(BaseModel):
    timelines: List[TimelineResponse] = []
    total: int = 0


class TimelineGenerateRequest(BaseModel):
    topic: Optional[str] = None
    collection_id: Optional[UUID] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
